metrics/request.go

```diff
@@ -0,0 +1,59 @@
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type RequestMetrics struct {
+	requestsTotal   *prometheus.CounterVec
+	requestDuration *prometheus.HistogramVec
+}
+
+func registerRequestMetrics(reg *prometheus.Registry) *RequestMetrics {
+	requestsTotal := registerCounterVec(reg, requestsMetricBaseName, []string{"method", "path", "status", "classification"})
+	requestDuration := registerDurationHistogram(reg, requestsMetricBaseName, []string{"method", "path"})
+
+	return &RequestMetrics{
+		requestsTotal:   requestsTotal,
+		requestDuration: requestDuration,
+	}
+}
+
+func (m *RequestMetrics) Track(method, path string) *RequestTracker {
+	return &RequestTracker{
+		method:  method,
+		path:    path,
+		metrics: m,
+	}
+}
+
+type RequestTracker struct {
+	method    string
+	path      string
+	startTime time.Time
+	metrics   *RequestMetrics
+}
+
+func (r *RequestTracker) Start() {
+	r.startTime = time.Now()
+}
+
+func (r *RequestTracker) End(status int) {
+	if status == 404 {
```

Comment:
But what it the endpoint is returning 404 not found because header is not found or something?

Reply:
I found a solution involving custom `gin.NoRoute` handler. So now the metrics-middleware can distinguish between "no-route" 404 from other 404s. 

If you don't like this new solution, I'm afraid that other option is to just delete this safeguard and accept such attack vector...